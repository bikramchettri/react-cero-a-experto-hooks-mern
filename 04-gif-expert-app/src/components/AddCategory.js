import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({setCategories}) => {
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
        console.log('handleInputChange trabajando');
    }
    const handleSubmit = (e) => {
        e.preventDefault();
        if(inputValue.trim().length > 2){
            setCategories( categorias => [inputValue, ...categorias]); 
            setInputValue('');
        }
    }
    return (
        <form onSubmit={handleSubmit}>
        <p>{inputValue}</p>
                <input
                type="text"
                value = {inputValue}
                onChange={ handleInputChange }
            />
        </form>
    )
}

AddCategory.propType = {
    setCategories: PropTypes.func.isRequired
}

