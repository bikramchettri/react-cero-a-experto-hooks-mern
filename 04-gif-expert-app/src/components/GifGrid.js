import React from 'react'
import  GifGridItem  from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';

export const GifGrid = ({category}) => {
    //const [images, setImages] = useState([]); //images el arreglo, setImages la funcion que lo llena
    const { data:images, loading } = useFetchGifs(category);
    
    return (
        <>
        <h3 className="animate__animated animate__flash">{category}</h3>
        {loading && <p>Cargando</p>}
             <div className="card-grid">
                
                {
                    images.map( img => ( 
                        <GifGridItem
                            key={img.id}
                            {...img}
                        />
                    ))
                }
            </div> 
        </>
    )
}
