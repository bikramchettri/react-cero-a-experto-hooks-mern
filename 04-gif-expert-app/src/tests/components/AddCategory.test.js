
import React from 'react';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom';
import {AddCategory} from '../../components/AddCategory';

describe('Pruebas en AddCategory', ()=> {
    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories} />);
    beforeEach( ()=>{
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories} />);

    });

    test('debe de mostrarse correctamente <AddCategory/>', async () => {

        expect(wrapper).toMatchSnapshot();
    });

    test('debe de cambiar la caja de texto', () => {
        
        const input = wrapper.find('input');

        const value = 'Hola mundo';

        input.simulate('change',{target : {value}});

        //const inputAfter = wrapper.find("input"); // en lugar de agregar el parrafo en el componente
        //expect(inputAfter.prop("value")).toBe(value); //obtenemos el nuevo valor del input

        expect(wrapper.find('p').text().trim()).toBe(value);
    });

    test('NO debe de postear la informacion obsubmit', () => {
        wrapper.find('form').simulate('submit', {preventDefault(){}});

        expect(setCategories).not.toHaveBeenCalled();

    });

    test('debe de llamar el setCategories y limpiar la caja de texto', () => {
        //1. simular el inputChange
        const value = 'Hola mundo';
        const input = wrapper.find('input');
        input.simulate('change',{target : {value}});
        //2.simular el submit
        wrapper.find('form').simulate('submit', {preventDefault(){}});
        //3. setCategories se debe de haber llamado
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1); //llamado una sola vez
        expect(setCategories).toHaveBeenCalledWith( expect.any(Function)); //se espera que se llame cualquier funcion

        //4. el valor del input debe ser ''
        expect(input.prop('value')).toBe('');

    })
    
    
    
})