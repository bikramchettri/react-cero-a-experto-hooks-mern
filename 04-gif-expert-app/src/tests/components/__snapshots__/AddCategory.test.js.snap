// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en AddCategory debe de mostrarse correctamente <AddCategory/> 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
